docs/tutorials/basic/data.md

```diff
@@ -192,17 +128,390 @@ class SimpleIter:
+You can learn more about the thinking behind the design of `RecordIO` [here](http://mxnet.io/architecture/note_data_loading.html).
+
+*MXNet* provides __`MXRecordIO`__ and __`MXIndexedRecordIO`__ for sequential access of data and random access of the data.
+
+### MXRecordIO
+First let's look at an example on how to read and write sequentially using `mx.recordio.MXRecordIO`. The files are named with a `.rec` extension.
+
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'w')
+for i in range(5):
+    record.write('record_%d'%i)
+record.close()
+```
+
+We can read the data back by opening the file with a option `r` as below:
+
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'r')
+while True:
+    item = record.read()
+    if not item:
+        break
+    print (item)
+record.close()
+```
+
+    record_0
+    record_1
+    record_2
+    record_3
+    record_4
+
+
+
+### MXIndexedRecordIO
+`MXIndexedRecordIO` supports random or indexed access to the data. We will create a indexed record file and a corresponding index file as below:
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'w')
+for i in range(5):
+    record.write_idx(i, 'record_%d'%i)
+record.close()
+```
+
+Now we can access the individual records using the keys
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'r')
+record.read_idx(3)
+```
+
+
+
+
+    'record_3'
+
+
+
+You can also list all the keys in the file.
+
+
+```python
+record.keys
+```
+
+
+
+
+    [0, 1, 2, 3, 4]
+
+
+
+
+### Packing and Unpacking data
+
+Each record in a .rec file can contain arbitrary binary data, however most deeplearning tasks need data in a  label/data format.  
+`mx.recordio` package provides a few utility functions for such operations, namely: `pack`, `unpack`, `pack_img`, and `unpack_img`.
+
+#### Packing/Unpacking Binary Data.
+
+[__`pack`__](http://mxnet.io/api/python/io.html#mxnet.recordio.pack) and [__`unpack`__](http://mxnet.io/api/python/io.html#mxnet.recordio.unpack) are used for storing float (or 1d array of float) label and binary data:
+
+
+```python
+# `pack`
+data = 'data'
+label1 = 1.0
+header1 = mx.recordio.IRHeader(flag=0, label=label1, id=1, id2=0)
+s1 = mx.recordio.pack(header1, data)
+print('float label:', repr(s1))
+label2 = [1.0, 2.0, 3.0]
+header2 = mx.recordio.IRHeader(flag=0, label=label2, id=2, id2=0)
+s2 = mx.recordio.pack(header2, data)
+print('array label:', repr(s2))
+```
 
-- `image_iter` by `ImageRecordIter` so we can enjoy the provided multithreaded
-pre-fetch and augmentation.
-- `caption_iter` by `NDArrayIter` or bucketing iterator provided in the `rnn`
-package.
+    ('float label:', "'\\x00\\x00\\x00\\x00\\x00\\x00\\x80?\\x01\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00data'")
+    ('array label:', "'\\x03\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x02\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x80?\\x00\\x00\\x00@\\x00\\x00@@data'")
```

Comment:
Do you see any use of printing s1 and s2 here. If not, feel free to remove that part.